python/ray/air/execution/_internal/event_manager.py

```diff
@@ -163,40 +234,452 @@ def wait(
+        - On task error, trigger actor task error callback
+        - On success, trigger actor task result callback
+        """
+        tracked_actor_task = self._futures_to_tracked_actor_tasks.pop(future)
+        tracked_actor = tracked_actor_task._tracked_actor
+        self._tracked_actors_to_futures[tracked_actor].remove(future)
+
+        try:
+            result = ray.get(future)
+        except RayActorError as exc:
+            # Here the actual actor process died.
+            # First, clean up any references to the actor and its futures
+            self._cleanup_failed_actor(tracked_actor)
+
+            # Actor failed - first trigger actor error callback
+            if tracked_actor._on_error:
+                tracked_actor._on_error(tracked_actor, exc)
+
+            # Then trigger actor task error callback
+            if tracked_actor_task._on_error:
+                tracked_actor_task._on_error(tracked_actor, exc)
+            return
+        except RayTaskError as exc:
+            # Here, only the task failed, but the actor process is still alive.
+            # Trigger actor task error callback
+            if tracked_actor_task._on_error:
+                tracked_actor_task._on_error(tracked_actor, exc)
+            return
+
+        # Trigger actor task result callback
+        if tracked_actor_task._on_result:
+            tracked_actor_task._on_result(tracked_actor, result)
+
+    def _handle_ready_resource_future(self):
+        """Handle a resource future that became ready.
+
+        - Update state of the resource manager
+        - Try to start one actor
+        """
+        # Force resource manager to update internal state
+        self._resource_manager.update_state()
+        # We handle resource futures one by one, so only try to start 1 actor at a time
+        self._try_start_actors(max_actors=1)
+
+    def _try_start_actors_timeout(
+        self, timeout: Union[int, float], max_actors: Optional[int] = None
+    ) -> Tuple[int, float]:
+        """Try starting up to ``max_actors`` actors in up to ``timeout`` seconds."""
+        now = start_time = time.monotonic()
+
+        started_actors = 0
+        while (
+            now < start_time + timeout
+            and self._tracked_actors_to_attrs
+            and not (
+                # If we have other futures to await, prefer that
+                self._futures_to_tracked_actor_tasks
+                or self._futures_to_actor_termination
+            )
+        ):
+            started_this_iter = self._try_start_actors(max_actors=max_actors)
+            started_actors += started_this_iter
+
+            if max_actors is not None:
+                max_actors -= started_this_iter
+                if max_actors <= 0:
+                    break
+
+            time.sleep(0.1)
+            now = time.monotonic()
+
+        return started_actors, now - start_time
+
+    def _try_start_actors(self, max_actors: Optional[int] = None) -> int:
```

Comment:
ok, at a high level, actor related things should live in ActorManager? which operates on top of resource and event manager?

Reply:
Yes! Updated